graphql_compiler/macros/macro_edge/validation.py

```diff
@@ -1,11 +1,102 @@
 from ...exceptions import GraphQLInvalidMacroError
-from .directives import MacroEdgeDefinitionDirective, MacroEdgeDirective, MacroEdgeTargetDirective
+from ...schema import VERTEX_FIELD_PREFIXES, is_vertex_field_name
+from .directives import MacroEdgeDefinitionDirective, MacroEdgeTargetDirective
 from .helpers import get_only_selection_from_ast
 
 
+def _validate_macro_ast_with_macro_directives(schema, ast, macro_directives):
+    """Raise errors if the macro uses the macro directives incorrectly or is otherwise invalid."""
+    if ast.directives is not None:
+        directive_names = [directive.name.value for directive in ast.directives]
+        raise GraphQLInvalidMacroError(
+            u'Unexpectedly found directives at the top level of the GraphQL input. '
+            u'This is not supported. Directives: {}'.format(directive_names))
+
+    if ast.variable_definitions is not None:
+        raise GraphQLInvalidMacroError(
+            u'Unexpectedly found variable definitions at the top level of the GraphQL input. '
+            u'This is not supported. Variable definitions: {}'.format(ast.variable_definitions))
+
+    required_macro_directives = (MacroEdgeDefinitionDirective, MacroEdgeTargetDirective)
+    schema_with_macro_directives = copy(schema)
+    schema_with_macro_directives._directives = list(chain(
+        schema_with_macro_directives._directives, required_macro_directives))
+
+    validation_errors = validate(schema, ast)
+    if validation_errors:
+        raise GraphQLInvalidMacroError(
+            u'Macro edge failed validation: {}'.format(validation_errors))
+
+    for directive_definition in required_macro_directives:
+        macro_data = macro_directives.get(directive_definition.name, None)
+        if not macro_data:
+            raise GraphQLInvalidMacroError(
+                u'Required macro edge directive "@{}" was not found anywhere within the supplied '
+                u'macro edge definition GraphQL.'.format(directive_definition.name))
+
+        if len(macro_data) > 1:
+            raise GraphQLInvalidMacroError(
+                u'Required macro edge directive "@{}" was unexpectedly present more than once in '
+                u'the supplied macro edge definition GraphQL. It was found {} times.'
+                .format(directive_definition.name, len(macro_data)))
+
+
+def _validate_class_selection_ast(ast, macro_defn_ast):
+    """Ensure that the macro's top-level selection AST adheres to our expectations."""
+    directive_names = [
+        directive.name.value
+        for directive in ast.directives
+    ]
+    unexpected_directives = [
+        directive_name
+        for directive_name in directive_names
+        if directive_name != MacroEdgeDefinitionDirective.name
+    ]
+    if unexpected_directives:
+        raise GraphQLInvalidMacroError(
+            u'Found unexpected directives at the top level of the macro definition GraphQL: '
+            u'{}'.format(unexpected_directives))
+
```

Comment:
Are you catching the case where the user puts two `@macro_edge_definition`s?

Reply:
Yes, that should raise near `if len(macro_data) > 1:` since that's the number of occurrences for a given macro edge directive.